### PR TITLE
feat: add hyperliquid core compatible interchain token

### PR DESCRIPTION
**Pull request by isi8787**

### CODE REVIEW

File: contracts/interchain-token/InterchainToken.sol [hunk 1]
```diff
@@ -15,10 +16,15 @@ import { Minter } from '../utils/Minter.sol';
...
  * @notice This contract implements an interchain token which extends InterchainToken functionality.
  * @dev This contract also inherits Minter and Implementation logic.
+ * Uses assembly to directly write to slot 0, bypassing parent contract storage layout.
  */
 contract InterchainToken is InterchainTokenStandard, ERC20, ERC20Permit, Minter, IInterchainToken {
     using AddressBytes for bytes;
 
+    /// @dev Storage slot for deployer address - FORCED to slot 0 using assembly
+    /// This is specifically slot 0 for Hyperliquid firstStorageSlot compatibility
+    bytes32 private constant DEPLOYER_SLOT = 0x0000000000000000000000000000000000000000000000000000000000000000;
```
Comment (milapsheth) [inline contracts/interchain-token/InterchainToken.sol:26]: The slot 0 needs to be reserved explicitly. Need to add a variable for it at the top.

Foivos recommended creating a separate HyperLiquidDeployer contract and inheriting from it first to do this cleanly

File: contracts/interchain-token/InterchainToken.sol [hunk 2]
```diff
@@ -15,10 +16,15 @@ import { Minter } from '../utils/Minter.sol';
  * @title InterchainToken
  * @notice This contract implements an interchain token which extends InterchainToken functionality.
  * @dev This contract also inherits Minter and Implementation logic.
+ * Uses assembly to directly write to slot 0, bypassing parent contract storage layout.
  */
 contract InterchainToken is InterchainTokenStandard, ERC20, ERC20Permit, Minter, IInterchainToken {
```
Comment (milapsheth) [inline contracts/interchain-token/InterchainToken.sol:21]: We need to maintain both the standard and new InterchainToken for Hyperliquid. Instead of forking the repo and having to maintain release processes for that, create a new HyperliquidInterchainToken that inherits from this contract and adds necessary modifications

File: contracts/interchain-token/InterchainToken.sol [hunk 3]
```diff
@@ -28,6 +34,41 @@ contract InterchainToken is InterchainTokenStandard, ERC20, ERC20Permit, Minter,
...
+        assembly {
+            sstore(DEPLOYER_SLOT, newDeployer)
+        }
+    }
+
+    /**
+     * @notice Allows the ITS contract or its operator to update the deployer address
+     * @param newDeployer The new deployer address to set
+     */
+    function updateDeployer(address newDeployer) external {
```
Comment (milapsheth) [inline contracts/interchain-token/InterchainToken.sol:63]: The ITS implementation will also need a method to call update on this

File: test/InterchainToken.js [hunk 4]
```diff
@@ -138,6 +140,21 @@ describe('InterchainToken', () => {
             const finalAllowance = await tokenTest.allowance(sender, spender);
             expect(finalAllowance).to.eq(initialAllowance);
```
Comment (milapsheth) [inline test/InterchainToken.js:141]: Can we add a test to ensure that the slot 0 is set and used correctly?

File: test/InterchainToken.js [hunk 5]
```diff
@@ -138,6 +140,21 @@ describe('InterchainToken', () => {
...
         });
```
Comment (milapsheth) [inline test/InterchainToken.js:142]: Deployment scripts need to be updated to support Hyperliquid deployment

File: contracts/interchain-token/HyperLiquidDeployer.sol [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+import { IOperator } from '../interfaces/IOperator.sol';
+
+/**
+ * @title HyperLiquidDeployer
+ * @notice This contract explicitly reserves slot 0 for the deployer address.
+ * @dev This is specifically for Hyperliquid firstStorageSlot compatibility.
+ * Must be inherited first to ensure slot 0 is properly reserved.
+ */
+abstract contract HyperLiquidDeployer {
```
Comment (milapsheth) [inline contracts/interchain-token/HyperLiquidDeployer.sol:13]:
```suggestion
abstract contract HyperliquidDeployer {
```
for consistency with other names

File: test/Slot0Reservation.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        // Slot 0 should contain deployer address
+        const slot0 = await provider.getStorageAt(tokenAddress, 0);
+        const deployerFromSlot0 = '0x' + slot0.slice(-40);
+        const deployerFromContract = await hyperliquidToken.getDeployer();
+
+        expect(deployerFromSlot0.toLowerCase()).to.equal(deployerFromContract.toLowerCase());
+    });
+
+    it('should verify slot 0 is reserved and not overwritten by other variables', async () => {
+        const provider = ethers.provider;
```
Comment (milapsheth) [inline test/Slot0Reservation.js:107]:
```suggestion
```
just import provider from ethers at the top? same below, instead of using fully qualified import names

File: package-lock.json [hunk 8]
```diff
@@ -1,12 +1,12 @@
...
   "name": "@axelar-network/interchain-token-service",
-  "version": "2.0.1",
+  "version": "2.1.0",
   "lockfileVersion": 3,
   "requires": true,
   "packages": {
     "": {
       "name": "@axelar-network/interchain-token-service",
-      "version": "2.0.1",
+      "version": "2.1.0",
```
Comment (milapsheth) [inline package-lock.json:9]: why this change? did you not branch off main? Looks like your PR has merge conflicts, can you resolve them?

File: test/ChainSpecificDeployment.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+        const deployment = await deployAll(
+            wallet,
+            'hyperliquid',
+            [],
+            'HyperliquidInterchainTokenService',
+            'HyperliquidInterchainTokenServiceFactory',
+        );
+
+        console.log('Deployed contracts:');
```
Comment (milapsheth) [inline test/ChainSpecificDeployment.js:25]: too much superfluous code in tests, please clean it up. harder to review, easy to miss the critical changes

File: test/ChainSpecificDeployment.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        const salt2 = ethers.utils.randomBytes(32);
+        const tokenId = ethers.utils.randomBytes(32);
+
+        await hyperliquidDeployment.activeTokenDeployer.deployInterchainToken(salt1, tokenId, wallet.address, 'TestToken', 'TEST', 18);
+        await standardDeployment.activeTokenDeployer.deployInterchainToken(salt2, tokenId, wallet.address, 'TestToken', 'TEST', 18);
+
+        const hyperliquidTokenAddress = await hyperliquidDeployment.activeTokenDeployer.deployedAddress(salt1);
+        const standardTokenAddress = await standardDeployment.activeTokenDeployer.deployedAddress(salt2);
+
+        // Check slot 0 for both tokens
```
Comment (milapsheth) [inline test/ChainSpecificDeployment.js:119]: there's already a slot test. why do we need this?

File: test/ChainSpecificDeployment.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        const provider = ethers.provider;
+        const hyperliquidSlot0 = await provider.getStorageAt(hyperliquidTokenAddress, 0);
+        const standardSlot0 = await provider.getStorageAt(standardTokenAddress, 0);
+
+        console.log('Storage slot 0 comparison:');
+        console.log('- Hyperliquid token slot 0:', hyperliquidSlot0);
+        console.log('- Standard token slot 0:', standardSlot0);
+
+        // Hyperliquid should have deployer in slot 0, standard should be empty
```
Comment (milapsheth) [inline test/ChainSpecificDeployment.js:128]: we don't care what the standard token does since there's no requirement of what slot 0 there should or should not be. Just need to test the hyperliquid one for which there's a separate test already

File: test/ChainSpecificDeployment.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
+'use strict';
+
+const { ethers } = require('hardhat');
+const { expect } = require('chai');
+const { deployAll } = require('../scripts/deploy');
+
+describe('Chain-Specific Token Deployment', () => {
```
Comment (milapsheth) [inline test/ChainSpecificDeployment.js:7]: don't see a point of this test

File: scripts/deploy.js [hunk 13]
```diff
@@ -91,8 +91,21 @@ async function deployAll(
 
     const interchainTokenServiceAddress = await getCreate3Address(create3Deployer.address, wallet, deploymentKey);
     const tokenManagerDeployer = await deployContract(wallet, 'TokenManagerDeployer', []);
+
+    // Deploy both token implementations
     const interchainToken = await deployContract(wallet, 'InterchainToken', [interchainTokenServiceAddress]);
+    const hyperliquidInterchainToken = await deployContract(wallet, 'HyperliquidInterchainToken', [interchainTokenServiceAddress]);
```
Comment (milapsheth) [inline scripts/deploy.js:97]: deploy the appropriate one based on whether chain is hyperliquid or not. deployment scripts will do the same

File: scripts/deploy.js [hunk 14]
```diff
@@ -91,8 +91,21 @@ async function deployAll(
...
+
+    // Deploy both token deployers
     const interchainTokenDeployer = await deployContract(wallet, 'InterchainTokenDeployer', [interchainToken.address]);
+    const hyperliquidInterchainTokenDeployer = await deployContract(wallet, 'HyperliquidInterchainTokenDeployer', [
```
Comment (milapsheth) [inline scripts/deploy.js:101]: this shouldn't be needed

File: test/HyperliquidInterchainToken.js [hunk 15]
```diff
@@ -0,0 +1,493 @@
...
+            const newDeployer = await token.getDeployer();
+            expect(newDeployer).to.equal(deployer.address);
+        });
+
+        it('should revert when non-ITS or non-operator tries to update deployer', async () => {
+            await expect(token.connect(user).updateDeployer(user.address)).to.be.reverted;
+        });
+
+        it('should store deployer in slot 0', async () => {
+            const provider = ethers.provider;
```
Comment (milapsheth) [inline test/HyperliquidInterchainToken.js:67]: import instead of doing `ethers.xyz` everywhere

File: test/HyperliquidInterchainToken.js [hunk 16]
```diff
@@ -0,0 +1,493 @@
...
+            const deployerFromContract = await token.getDeployer();
+
+            // Both should match
+            expect(deployerFromSlot0.toLowerCase()).to.equal(deployerFromContract.toLowerCase());
+
+            // Slot 0 should not be empty
+            expect(slot0).to.not.equal('0x0000000000000000000000000000000000000000000000000000000000000000');
+        });
+
+        it('should maintain ERC20 functionality', async () => {
```
Comment (milapsheth) [inline test/HyperliquidInterchainToken.js:84]: don't need to re-test token functionality. since the contract is inheriting existing functionality, we can assume it works. If we really need to retest some logic, the same test should be run on the two separate token instances instead of duplicating tests

File: test/HyperliquidInterchainToken.js [hunk 17]
```diff
@@ -0,0 +1,493 @@
...
+            // Call updateDeployer as ITS (owner)
+            const newDeployer = user.address;
+            await testDeployer.connect(owner).updateDeployer(newDeployer);
+
+            // Check that the deployer was updated
+            expect(await testDeployer.getDeployer()).to.equal(newDeployer);
+        });
+    });
+
+    describe('Storage Layout Verification', () => {
```
Comment (milapsheth) [inline test/HyperliquidInterchainToken.js:146]: there's a separate slot 0 test file? See comments there

File: test/HyperliquidInterchainToken.js [hunk 18]
```diff
@@ -0,0 +1,493 @@
...
+        it('Should preserve the same bytecode', async () => {
+            const contract = await ethers.getContractFactory('HyperliquidInterchainToken', owner);
+            const contractBytecode = contract.bytecode;
+
+            // For now, just verify the bytecode is not empty
+            expect(contractBytecode).to.not.be.empty;
+        });
+    });
+
+    describe('Coverage Improvement Tests', () => {
```
Comment (milapsheth) [inline test/HyperliquidInterchainToken.js:183]: vague grouping. see our naming and grouping pattern in other test files

File: test/HyperliquidInterchainToken.js [hunk 19]
```diff
@@ -0,0 +1,493 @@
...
+        it('should test HyperLiquidDeployer getDeployer assembly code', async () => {
```
Comment (milapsheth) [inline test/HyperliquidInterchainToken.js:184]: tests should describe the usage, not implementation details, see our existing tests
```suggestion
        it('should get the deployer', async () => {
```

File: test/HyperliquidInterchainToken.js [hunk 20]
```diff
@@ -0,0 +1,493 @@
...
+            const deployer = await token.getDeployer();
+            expect(deployer).to.not.equal(ethers.constants.AddressZero);
+        });
+
+        it('should test HyperLiquidDeployer updateDeployer authorization', async () => {
```
Comment (milapsheth) [inline test/HyperliquidInterchainToken.js:189]:
```suggestion
        it('should update the deployer', async () => {
```

File: test/HyperliquidInterchainToken.js [hunk 21]
```diff
@@ -0,0 +1,493 @@
...
+            const newDeployer = user.address;
+            await token.connect(owner).updateDeployer(newDeployer);
+            expect(await token.getDeployer()).to.equal(newDeployer);
+        });
+
+        it('should test HyperLiquidDeployer updateDeployer failure case', async () => {
```
Comment (milapsheth) [inline test/HyperliquidInterchainToken.js:195]:
```suggestion
        it('should revert on updating deployer from non-operator address', async () => {
```

File: test/HyperliquidInterchainToken.js [hunk 22]
```diff
@@ -0,0 +1,493 @@
...
+            await token.connect(owner).updateDeployer(ethers.constants.AddressZero);
+            const slot0Zero = await provider.getStorageAt(tokenAddress, 0);
+            const deployerFromSlot0Zero = '0x' + slot0Zero.slice(-40);
+            const deployerFromContractZero = await token.getDeployer();
+
+            expect(deployerFromSlot0Zero.toLowerCase()).to.equal(deployerFromContractZero.toLowerCase());
+            expect(deployerFromContractZero).to.equal(ethers.constants.AddressZero);
+        });
+
+        it('should test HyperLiquidDeployer assembly code with edge cases', async () => {
```
Comment (milapsheth) [inline test/HyperliquidInterchainToken.js:333]: doesn't seem very useful. just set and test a random address. if you need to have test vectors, the test should loop over the vectors, instead of duplicating code per test

File: test/HyperliquidInterchainToken.js [hunk 23]
```diff
@@ -0,0 +1,493 @@
...
+            expect(returnedAlternatingAddress.toLowerCase()).to.equal(alternatingAddress.toLowerCase());
+
+            // Verify storage slot
+            const slot0Alt = await provider.getStorageAt(tokenAddress, 0);
+            const deployerFromSlot0Alt = '0x' + slot0Alt.slice(-40);
+            expect(deployerFromSlot0Alt.toLowerCase()).to.equal(returnedAlternatingAddress.toLowerCase());
+        });
+
+        it('should test HyperLiquidDeployer _setDeployer assembly code with multiple rapid updates', async () => {
+            // Test rapid updates to ensure _setDeployer assembly code is thoroughly covered
```
Comment (milapsheth) [inline test/HyperliquidInterchainToken.js:373]: what is this trying to test? stress testing the local hardhat EVM lol?

File: contracts/InterchainTokenService.sol
Comment (milapsheth) [inline contracts/InterchainTokenService.sol:1]: PR name is not very descriptive. Maybe something like
`feat: add hyperliquid core compatible interchain token`

Comment (milapsheth) [reply contracts/InterchainTokenService.sol:1]: run `npm run cs` to add a changelog for a minor version bump

File: contracts/interchain-token/HyperliquidInterchainToken.sol [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+     */
+    function _getInterchainTokenService() internal view override returns (address) {
+        return interchainTokenService();
+    }
+
+    /**
+     * @notice Override interchainTokenId function from InterchainToken
+     * @return bytes32 The tokenId of the token
+     */
+    function interchainTokenId() public view override(InterchainToken, IInterchainToken) returns (bytes32) {
```
[the remark this answers is not in the document]
Comment (isi8787) [reply contracts/interchain-token/HyperliquidInterchainToken.sol:81]: refactoring to inherit directly from InterchainToken so we dont redo this code

File: contracts/interchain-token/HyperliquidInterchainToken.sol [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+        return super.interchainTokenService();
+    }
+
+    function getDeployer() external view override(HyperLiquidDeployer, IHyperliquidInterchainToken) returns (address deployer) {
+        assembly {
+            deployer := sload(0)
+        }
+    }
+
+    function updateDeployer(address newDeployer) external override(HyperLiquidDeployer, IHyperliquidInterchainToken) {
```
[the remark this answers is not in the document]
Comment (isi8787) [reply contracts/interchain-token/HyperliquidInterchainToken.sol:99]: removed from HyperliquidDeployer.sol to keep logic seperate

File: contracts/interchain-token/HyperLiquidDeployer.sol [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+    /**
+     * @notice Allows the ITS contract or its operator to update the deployer address
+     * @param newDeployer The new deployer address to set
+     */
+    function updateDeployer(address newDeployer) external virtual {
+        address its = _getInterchainTokenService();
+        if (msg.sender != its) {
+            // Check if caller is ITS operator
+            if (!IOperator(its).isOperator(msg.sender)) revert NotAuthorized();
+        }
```
[the remark this answers is not in the document]
Comment (isi8787) [reply contracts/interchain-token/HyperLiquidDeployer.sol:51]: im going to add a service that can abstract away this level of coupling, but the intention is to still have control over who can set the deployer address

File: contracts/InterchainTokenService.sol [hunk 27]
```diff
@@ -658,6 +659,27 @@ contract InterchainTokenService is
...
+     */
+    function updateHyperliquidTokenDeployer(address tokenAddress, address newDeployer) external onlyRole(uint8(Roles.OPERATOR)) {
+        if (tokenAddress == address(0)) revert ZeroAddress();
+        if (newDeployer == address(0)) revert ZeroAddress();
+
+        // Try to call updateDeployer on the token
+        // This will work if the token is a HyperliquidInterchainToken
+        // and will revert if it's not (which is the desired behavior)
+        try IHyperliquidInterchainToken(tokenAddress).updateDeployer(newDeployer) {
+            // Success - the token was a HyperliquidInterchainToken
```
[the remark this answers is not in the document]
Comment (isi8787) [reply contracts/InterchainTokenService.sol:676]: lol; yes cleaning  up

File: contracts/interchain-token/HyperLiquidDeployer.sol [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+        _setDeployer(newDeployer);
+    }
+
+    /**
+     * @notice Abstract function to get the interchain token service address
+     * @return address The interchain token service contract address
+     */
+    function _getInterchainTokenService() internal view virtual returns (address);
```
[the remark this answers is not in the document]
Comment (isi8787) [reply contracts/interchain-token/HyperLiquidDeployer.sol:59]: removed

[the remark this answers is not in the document]
Comment (isi8787) [reply contracts/interchain-token/HyperLiquidDeployer.sol:59]: pushing a modified version where i am moving the logic to HyperliquidInterchainToken.sol

File: test/Slot0Reservation.js [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+
+        // Update deployer (only owner can do this as they are the ITS)
+        await hyperliquidToken.connect(owner).updateDeployer(user.address);
+
+        // Read updated slot 0
+        const updatedSlot0 = await provider.getStorageAt(tokenAddress, 0);
+        const updatedDeployer = await hyperliquidToken.getDeployer();
+
+        // Verify changes
+        expect(updatedDeployer.toLowerCase()).to.equal(user.address.toLowerCase());
```
[the remark this answers is not in the document]
Comment (isi8787) [reply test/Slot0Reservation.js]: done

File: test/Slot0Reservation.js [hunk 30]
```diff
@@ -0,0 +1,144 @@
...
+            const isEmpty = slot === '0x0000000000000000000000000000000000000000000000000000000000000000';
+            const asAddress = '0x' + slot.slice(-40);
+
+            console.log(`Slot ${i}: ${slot} ${isEmpty ? '(empty)' : ''}`);
+
+            if (!isEmpty) {
+                console.log(`  As address: ${asAddress}`);
+                console.log(`  As number: ${parseInt(slot, 16)}`);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (isi8787) [reply test/Slot0Reservation.js:96]: taking care to eliminate all LLM fluff

File: test/Slot0Reservation.js [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+
+        // Slot 0 should contain deployer address
+        const slot0 = await provider.getStorageAt(tokenAddress, 0);
+        const deployerFromSlot0 = '0x' + slot0.slice(-40);
+        const deployerFromContract = await hyperliquidToken.getDeployer();
+
+        expect(deployerFromSlot0.toLowerCase()).to.equal(deployerFromContract.toLowerCase());
```
[the remark this answers is not in the document]
Comment (isi8787) [reply test/Slot0Reservation.js:103]: done

File: test/HyperliquidInterchainToken.js (hunk 17 above)
Comment (isi8787) [reply test/HyperliquidInterchainToken.js:146]: removing duplicated test
